ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/storage/flat/FlatDbReaderStrategy.java

```diff
@@ -0,0 +1,175 @@
+
+  public FlatDbReaderStrategy(
+      final MetricsSystem metricsSystem,
+      final KeyValueStorage accountStorage,
+      final KeyValueStorage codeStorage,
+      final KeyValueStorage storageStorage) {
+    this.metricsSystem = metricsSystem;
+    this.accountStorage = accountStorage;
+    this.codeStorage = codeStorage;
+    this.storageStorage = storageStorage;
+
+    getAccountCounter =
+        metricsSystem.createCounter(
+            BesuMetricCategory.BLOCKCHAIN,
+            "get_account_total",
+            "Total number of calls to getAccount");
+
+    getAccountFlatDatabaseCounter =
+        metricsSystem.createCounter(
+            BesuMetricCategory.BLOCKCHAIN,
+            "get_account_flat_database",
+            "Number of accounts found in the flat database");
+
+    getStorageValueCounter =
+        metricsSystem.createCounter(
+            BesuMetricCategory.BLOCKCHAIN,
+            "get_storagevalue_total",
+            "Total number of calls to getStorageValueBySlotHash");
+
+    getStorageValueFlatDatabaseCounter =
+        metricsSystem.createCounter(
+            BesuMetricCategory.BLOCKCHAIN,
+            "get_storagevalue_flat_database",
+            "Number of storage slots found in the flat database");
+  }
+
+  /*
+   * Retrieves the account data for the given account hash, using the world state root hash supplier and node loader.
+   */
+  public Optional<Bytes> getAccount(
+      final Supplier<Optional<Bytes>> worldStateRootHashSupplier,
+      final NodeLoader nodeLoader,
+      final Hash accountHash) {
+    return accountStorage.get(accountHash.toArrayUnsafe()).map(Bytes::wrap);
+  }
+
+  /*
+   * Retrieves the code data for the given code hash and account hash.
+   */
+  public Optional<Bytes> getCode(final Bytes32 codeHash, final Hash accountHash) {
+    if (codeHash.equals(Hash.EMPTY)) {
+      return Optional.of(Bytes.EMPTY);
+    } else {
+      return codeStorage
+          .get(accountHash.toArrayUnsafe())
+          .map(Bytes::wrap)
+          .filter(b -> Hash.hash(b).equals(codeHash));
+    }
+  }
+
+  /*
+   * Retrieves the storage value for the given account hash and storage slot key, using the world state root hash supplier, storage root supplier, and node loader.
+   */
+  public Optional<Bytes> getStorageValueByStorageSlotKey(
+      final Supplier<Optional<Bytes>> worldStateRootHashSupplier,
+      final Supplier<Optional<Hash>> storageRootSupplier,
+      final NodeLoader nodeLoader,
+      final Hash accountHash,
+      final StorageSlotKey storageSlotKey) {
+    return storageStorage
+        .get(Bytes.concatenate(accountHash, storageSlotKey.getSlotHash()).toArrayUnsafe())
+        .map(Bytes::wrap);
+  }
+
+  public void clearAll() {
+    accountStorage.clear();
+    storageStorage.clear();
+    codeStorage.clear();
+  }
+
+  public void clearAccountAndStorageDatabase() {
+    accountStorage.clear();
+    storageStorage.clear();
+  }
```

Comment:
when calling debug_resyncWorldState on a FULL db, this is fine.  However when calling on a PARTIAL db, I get a nasty stack trace from the download pipeline

this isn't the right place to dump this, but it looks like some tasks in the pipeline are not getting notified of the truncation and the pipeline abends noisily:
```
rocketpool_eth1  | FLAT DB MODE IS PARTIAL org.hyperledger.besu.ethereum.bonsai.storage.flat.PartialFlatDbReaderStrategy@cae295f5
rocketpool_eth1  | 2023-05-24 23:41:32.178+00:00 | EthScheduler-Services-3 (batchPersistAccountData) | INFO  | Pipeline | Unhandled exception in pipeline. Aborting.
rocketpool_eth1  | org.hyperledger.besu.plugin.services.exception.StorageException: org.rocksdb.RocksDBException: Could not access column family 16733
rocketpool_eth1  |      at org.hyperledger.besu.plugin.services.storage.rocksdb.segmented.RocksDBColumnarKeyValueStorage$RocksDbTransaction.commit(RocksDBColumnarKeyValueStorage.java:397)
rocketpool_eth1  |      at org.hyperledger.besu.services.kvstore.SegmentedKeyValueStorageTransactionTransitionValidatorDecorator.commit(SegmentedKeyValueStorageTransactionTransitionValidatorDecorator.java:59)
rocketpool_eth1  |      at org.hyperledger.besu.services.kvstore.SegmentedKeyValueStorageAdapter$1.commit(SegmentedKeyValueStorageAdapter.java:119)
rocketpool_eth1  |      at org.hyperledger.besu.ethereum.bonsai.storage.BonsaiWorldStateKeyValueStorage$Updater.commit(BonsaiWorldStateKeyValueStorage.java:502)
rocketpool_eth1  |      at org.hyperledger.besu.ethereum.eth.sync.snapsync.PersistDataStep.persist(PersistDataStep.java:80)
rocketpool_eth1  |      at org.hyperledger.besu.ethereum.eth.sync.snapsync.PersistDataStep.persist(PersistDataStep.java:107)
rocketpool_eth1  |      at org.hyperledger.besu.ethereum.eth.sync.snapsync.SnapWorldStateDownloadProcess$Builder.lambda$build$4(SnapWorldStateDownloadProcess.java:262)
rocketpool_eth1  |      at org.hyperledger.besu.services.pipeline.MapProcessor.processNextInput(MapProcessor.java:31)
rocketpool_eth1  |      at org.hyperledger.besu.services.pipeline.ProcessingStage.run(ProcessingStage.java:38)
rocketpool_eth1  |      at org.hyperledger.besu.services.pipeline.Pipeline.lambda$runWithErrorHandling$3(Pipeline.java:167)
rocketpool_eth1  |      at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
rocketpool_eth1  |      at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
rocketpool_eth1  |      at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
rocketpool_eth1  |      at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
rocketpool_eth1  |      at java.base/java.lang.Thread.run(Thread.java:857)
rocketpool_eth1  | Caused by: org.rocksdb.RocksDBException: Could not access column family 16733
rocketpool_eth1  |      at org.rocksdb.Transaction.commit(Native Method)
rocketpool_eth1  |      at org.rocksdb.Transaction.commit(Transaction.java:208)
rocketpool_eth1  |      at org.hyperledger.besu.plugin.services.storage.rocksdb.segmented.RocksDBColumnarKeyValueStorage$RocksDbTransaction.commit(RocksDBColumnarKeyValueStorage.java:391)
rocketpool_eth1  |      ... 14 more
rocketpool_eth1  | 2023-05-24 23:41:32.193+00:00 | EthScheduler-Services-2 (batchDownloadAccountData) | INFO  | Pipeline | Unhandled exception in pipeline. Aborting.
rocketpool_eth1  | java.lang.RuntimeException: Async operation failed. java.util.concurrent.CancellationException
rocketpool_eth1  |      at org.hyperledger.besu.services.pipeline.AsyncOperationProcessor.outputNextCompletedTask(AsyncOperationProcessor.java:86)
rocketpool_eth1  |      at org.hyperledger.besu.services.pipeline.AsyncOperationProcessor.attemptFinalization(AsyncOperationProcessor.java:70)
rocketpool_eth1  |      at org.hyperledger.besu.services.pipeline.ProcessingStage.run(ProcessingStage.java:43)
rocketpool_eth1  |      at org.hyperledger.besu.services.pipeline.Pipeline.lambda$runWithErrorHandling$3(Pipeline.java:167)
rocketpool_eth1  |      at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
rocketpool_eth1  |      at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
rocketpool_eth1  |      at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
rocketpool_eth1  |      at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
rocketpool_eth1  |      at java.base/java.lang.Thread.run(Thread.java:857)
rocketpool_eth1  | Caused by: java.util.concurrent.ExecutionException: java.util.concurrent.CancellationException
rocketpool_eth1  |      at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
rocketpool_eth1  |      at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2096)
rocketpool_eth1  |      at org.hyperledger.besu.services.pipeline.AsyncOperationProcessor.waitForAnyFutureToComplete(AsyncOperationProcessor.java:94)
rocketpool_eth1  |      at org.hyperledger.besu.services.pipeline.AsyncOperationProcessor.outputNextCompletedTask(AsyncOperationProcessor.java:81)
rocketpool_eth1  |      ... 8 more
rocketpool_eth1  | Caused by: java.util.concurrent.CancellationException
rocketpool_eth1  |      at java.base/java.util.concurrent.CompletableFuture.cancel(CompletableFuture.java:2478)
rocketpool_eth1  |      at org.hyperledger.besu.services.pipeline.AsyncOperationProcessor.lambda$abort$1(AsyncOperationProcessor.java:76)
rocketpool_eth1  |      at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
rocketpool_eth1  |      at org.hyperledger.besu.services.pipeline.AsyncOperationProcessor.abort(AsyncOperationProcessor.java:76)
rocketpool_eth1  |      at org.hyperledger.besu.services.pipeline.ProcessingStage.run(ProcessingStage.java:41)
rocketpool_eth1  |      ... 6 more
rocketpool_eth1  | 2023-05-24 23:41:32.538+00:00 | EthScheduler-Services-8 (batchPersistStorageData) | INFO  | Pipeline | Unhandled exception in pipeline. Aborting.
rocketpool_eth1  | org.hyperledger.besu.plugin.services.exception.StorageException: org.rocksdb.RocksDBException: Default cf timestamp size mismatch
rocketpool_eth1  |      at org.hyperledger.besu.plugin.services.storage.rocksdb.segmented.RocksDBColumnarKeyValueStorage$RocksDbTransaction.put(RocksDBColumnarKeyValueStorage.java:371)
rocketpool_eth1  |      at org.hyperledger.besu.plugin.services.storage.rocksdb.segmented.RocksDBColumnarKeyValueStorage$RocksDbTransaction.put(RocksDBColumnarKeyValueStorage.java:346)
rocketpool_eth1  |      at org.hyperledger.besu.services.kvstore.SegmentedKeyValueStorageTransactionTransitionValidatorDecorator.put(SegmentedKeyValueStorageTransactionTransitionValidatorDecorator.java:46)
rocketpool_eth1  |      at org.hyperledger.besu.services.kvstore.SegmentedKeyValueStorageAdapter$1.put(SegmentedKeyValueStorageAdapter.java:109)
rocketpool_eth1  |      at org.hyperledger.besu.ethereum.bonsai.storage.BonsaiWorldStateKeyValueStorage$Updater.putStorageValueBySlotHash(BonsaiWorldStateKeyValueStorage.java:479)
rocketpool_eth1  |      at org.hyperledger.besu.ethereum.eth.sync.snapsync.request.StorageRangeDataRequest.lambda$doPersist$1(StorageRangeDataRequest.java:108)
rocketpool_eth1  |      at java.base/java.util.TreeMap.forEach(TreeMap.java:1282)
rocketpool_eth1  |      at org.hyperledger.besu.ethereum.eth.sync.snapsync.StackTrie.commit(StackTrie.java:130)
rocketpool_eth1  |      at org.hyperledger.besu.ethereum.eth.sync.snapsync.request.StorageRangeDataRequest.doPersist(StorageRangeDataRequest.java:112)
rocketpool_eth1  |      at org.hyperledger.besu.ethereum.eth.sync.snapsync.request.SnapDataRequest.persist(SnapDataRequest.java:123)
rocketpool_eth1  |      at org.hyperledger.besu.ethereum.eth.sync.snapsync.PersistDataStep.persist(PersistDataStep.java:65)
rocketpool_eth1  |      at org.hyperledger.besu.ethereum.eth.sync.snapsync.SnapWorldStateDownloadProcess$Builder.lambda$build$8(SnapWorldStateDownloadProcess.java:285)
rocketpool_eth1  |      at org.hyperledger.besu.services.pipeline.MapProcessor.processNextInput(MapProcessor.java:31)
rocketpool_eth1  |      at org.hyperledger.besu.services.pipeline.ProcessingStage.run(ProcessingStage.java:38)
rocketpool_eth1  |      at org.hyperledger.besu.services.pipeline.Pipeline.lambda$runWithErrorHandling$3(Pipeline.java:167)
rocketpool_eth1  |      at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
rocketpool_eth1  |      at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
rocketpool_eth1  |      at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
rocketpool_eth1  |      at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
rocketpool_eth1  |      at java.base/java.lang.Thread.run(Thread.java:857)
rocketpool_eth1  | Caused by: org.rocksdb.RocksDBException: Default cf timestamp size mismatch
rocketpool_eth1  |      at org.rocksdb.Transaction.put(Native Method)
rocketpool_eth1  |      at org.rocksdb.Transaction.put(Transaction.java:857)
rocketpool_eth1  |      at org.hyperledger.besu.plugin.services.storage.rocksdb.segmented.RocksDBColumnarKeyValueStorage$RocksDbTransaction.put(RocksDBColumnarKeyValueStorage.java:365)
rocketpool_eth1  |      ... 19 more
rocketpool_eth1  | 2023-05-24 23:41:32.540+00:00 | EthScheduler-Services-7 (batchDownloadStorageData) | INFO  | Pipeline | Unhandled exception in pipeline. Aborting.
rocketpool_eth1  | java.util.concurrent.CancellationException
rocketpool_eth1  |      at java.base/java.util.concurrent.CompletableFuture.cancel(CompletableFuture.java:2478)
rocketpool_eth1  |      at org.hyperledger.besu.services.pipeline.AsyncOperationProcessor.lambda$abort$1(AsyncOperationProcessor.java:76)
rocketpool_eth1  |      at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
rocketpool_eth1  |      at org.hyperledger.besu.services.pipeline.AsyncOperationProcessor.abort(AsyncOperationProcessor.java:76)
rocketpool_eth1  |      at org.hyperledger.besu.services.pipeline.ProcessingStage.run(ProcessingStage.java:41)
rocketpool_eth1  |      at org.hyperledger.besu.services.pipeline.Pipeline.lambda$runWithErrorHandling$3(Pipeline.java:167)
rocketpool_eth1  |      at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
rocketpool_eth1  |      at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
rocketpool_eth1  |      at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
rocketpool_eth1  |      at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
rocketpool_eth1  |      at java.base/java.lang.Thread.run(Thread.java:857)
```

Reply:
Yes, I think there was a snap sync running at the time.